File: roblox_studio_physical_operation_mcp/toolbar_detector.py
"""
工具栏按钮状态检测模块 - 模板匹配版本

使用 OpenCV 模板匹配在截图中定位按钮，然后通过颜色判断状态。

按钮布局（从左到右）：
1. 播放按钮(Play): 三角形 ▶
2. 暂停按钮(Pause): 双竖线 ‖  
3. 停止按钮(Stop): 方块 ■

颜色状态：
- 灰色 = disabled
- 彩色 = enabled (绿色播放、蓝色暂停、红色停止)
"""

import os
from typing import Optional, List, Tuple
from dataclasses import dataclass
from enum import Enum

try:
    import cv2
    import numpy as np
    from PIL import Image
except ImportError as e:
    raise ImportError(f"需要安装依赖: pip install opencv-python Pillow numpy: {e}")


class ButtonState(Enum):
    """按钮状态"""
    DISABLED = "disabled"  # 灰色，不可用
    ENABLED = "enabled"    # 彩色，可用
    UNKNOWN = "unknown"    # 无法识别


class ButtonType(Enum):
    """按钮类型"""
    PLAY = "play"
    PAUSE = "pause"
    STOP = "stop"


@dataclass
class ButtonInfo:
    """按钮信息"""
    button_type: ButtonType
    x: int
    y: int
    width: int
    height: int
    state: ButtonState
    color_type: str  # "gray", "green", "blue", "red"
    confidence: float  # 模板匹配置信度


@dataclass 
class ToolbarState:
    """工具栏状态"""
    play: ButtonState
    pause: ButtonState
    stop: ButtonState
    game_state: str  # "stopped", "running", "paused"
    buttons: List[ButtonInfo] = None
    
    def to_dict(self) -> dict:
        result = {
            "play": self.play.value,
            "pause": self.pause.value,
            "stop": self.stop.value,
            "game_state": self.game_state
        }
        if self.buttons:
            result["buttons_detail"] = [
                {
                    "type": b.button_type.value,
                    "x": b.x, "y": b.y,
                    "width": b.width, "height": b.height,
                    "state": b.state.value,
                    "color_type": b.color_type,
                    "confidence": round(b.confidence, 3)
                }
                for b in self.buttons
            ]
        return result


# 模板文件路径
TEMPLATE_DIR = os.path.join(os.path.dirname(__file__), "templates")


def capture_window_to_image(hwnd: int) -> Optional[Image.Image]:
    """截取窗口为 PIL Image 对象"""
    from .platform_utils import capture_window_to_pil
    return capture_window_to_pil(hwnd)


def pil_to_cv2(pil_image: Image.Image) -> np.ndarray:
    """PIL Image 转 OpenCV 格式"""
    return cv2.cvtColor(np.array(pil_image), cv2.COLOR_RGB2BGR)


def rgb_to_hsv(r: int, g: int, b: int) -> Tuple[float, float, float]:
    """RGB 转 HSV (H: 0-360, S: 0-100, V: 0-100)"""
    r, g, b = r / 255.0, g / 255.0, b / 255.0
    max_val = max(r, g, b)
    min_val = min(r, g, b)
    diff = max_val - min_val
    
    v = max_val * 100
    
    if max_val == 0:
        s = 0
    else:
        s = (diff / max_val) * 100
    
    if diff == 0:
        h = 0
    elif max_val == r:
        h = 60 * (((g - b) / diff) % 6)
    elif max_val == g:
        h = 60 * (((b - r) / diff) + 2)
    else:
        h = 60 * (((r - g) / diff) + 4)
    
    if h < 0:
        h += 360
        
    return h, s, v


def load_template(name: str) -> Optional[np.ndarray]:
    """加载模板图片（灰度）"""
    path = os.path.join(TEMPLATE_DIR, f"{name}.png")
    if not os.path.exists(path):
        return None
    return cv2.imread(path, cv2.IMREAD_GRAYSCALE)


def find_button_by_template(screenshot_gray: np.ndarray, template: np.ndarray, 
                            threshold: float = 0.7) -> Optional[Tuple[int, int, float]]:
    """
    使用模板匹配在截图中查找按钮
    
    Args:
        screenshot_gray: 灰度截图
        template: 灰度模板
        threshold: 匹配阈值
        
    Returns:
        (x, y, confidence) 或 None
    """
    if template is None:
        return None
    
    result = cv2.matchTemplate(screenshot_gray, template, cv2.TM_CCOEFF_NORMED)
    min_val, max_val, min_loc, max_loc = cv2.minMaxLoc(result)
    
    if max_val >= threshold:
        return (max_loc[0], max_loc[1], max_val)
    
    return None


def analyze_button_color(pil_image: Image.Image, x: int, y: int, 
                         width: int, height: int, button_type: str = None) -> Tuple[ButtonState, str]:
    """
    分析按钮区域的颜色来判断状态
    
    对于暂停按钮，使用亮度判断（竖线颜色）：
    - 浅灰色 (亮度 > 120) = disabled (游戏停止)
    - 深灰色 (亮度 < 120) = enabled (游戏运行中)
    
    对于其他按钮，使用彩色判断：
    - 有彩色 = enabled
    - 灰色 = disabled
    
    Returns:
        (state, color_type)
    """
    # 饱和度阈值
    COLOR_THRESHOLD = 50
    
    red_count = 0
    green_count = 0
    blue_count = 0
    dark_pixel_count = 0  # 深色像素计数
    total = 0
    total_dark_brightness = 0  # 深色像素的亮度总和
    
    for dy in range(height):
        for dx in range(width):
            try:
                r, g, b = pil_image.getpixel((x + dx, y + dy))[:3]
                h, s, v = rgb_to_hsv(r, g, b)
                brightness = (r + g + b) / 3
                total += 1
                
                # 统计非白色像素（按钮图标本身）
                if brightness < 200:
                    dark_pixel_count += 1
                    total_dark_brightness += brightness
                
                if s > COLOR_THRESHOLD and v > 30:
                    if h <= 30 or h >= 330:  # 红色
                        red_count += 1
                    elif 80 <= h <= 160:  # 绿色
                        green_count += 1
                    elif 180 <= h <= 250:  # 蓝色
                        blue_count += 1
            except:
                pass
    
    if total == 0:
        return ButtonState.UNKNOWN, "unknown"
    
    # 暂停按钮特殊处理：用图标亮度判断
    if button_type == "pause" and dark_pixel_count > 0:
        avg_dark_brightness = total_dark_brightness / dark_pixel_count
        # 亮度 > 120 = 浅灰色 = disabled (游戏停止)
        # 亮度 < 120 = 深灰色 = enabled (游戏运行中)
        if avg_dark_brightness < 120:
            return ButtonState.ENABLED, "dark"
        else:
            return ButtonState.DISABLED, "light"
    
    # 其他按钮：用彩色判断
    min_ratio = 0.03
    if red_count > total * min_ratio:
        return ButtonState.ENABLED, "red"
    elif green_count > total * min_ratio:
        return ButtonState.ENABLED, "green"
    elif blue_count > total * min_ratio:
        return ButtonState.ENABLED, "blue"
    else:
        return ButtonState.DISABLED, "gray"


def restore_window_if_minimized(hwnd: int) -> bool:
    """
    如果窗口最小化，则恢复窗口

    Args:
        hwnd: 窗口句柄

    Returns:
        True 如果窗口已恢复或本来就不是最小化
    """
    from .platform_utils import restore_window_if_minimized as _restore
    return _restore(hwnd)


def detect_toolbar_state(hwnd: int) -> Optional[ToolbarState]:
    """
    检测工具栏按钮状态
    
    Args:
        hwnd: Roblox Studio 窗口句柄
        
    Returns:
        ToolbarState 对象，失败返回 None
    """
    # 如果窗口最小化，先恢复
    restore_window_if_minimized(hwnd)
    
    # 截取窗口
    pil_image = capture_window_to_image(hwnd)
    if pil_image is None:
        return None
    
    # 转换为 OpenCV 格式
    cv_image = pil_to_cv2(pil_image)
    gray_image = cv2.cvtColor(cv_image, cv2.COLOR_BGR2GRAY)
    
    # 加载模板
    play_template = load_template("play")
    pause_template = load_template("pause")
    stop_template = load_template("stop")
    
    buttons = []
    play_state = ButtonState.UNKNOWN
    pause_state = ButtonState.UNKNOWN
    stop_state = ButtonState.UNKNOWN
    play_color = None
    stop_color = None
    
    # 查找播放按钮
    if play_template is not None:
        result = find_button_by_template(gray_image, play_template)
        if result:
            x, y, confidence = result
            h, w = play_template.shape
            state, color = analyze_button_color(pil_image, x, y, w, h)
            play_state = state
            play_color = color
            buttons.append(ButtonInfo(
                button_type=ButtonType.PLAY,
                x=x, y=y, width=w, height=h,
                state=state, color_type=color, confidence=confidence
            ))
    
    # 查找暂停按钮
    if pause_template is not None:
        result = find_button_by_template(gray_image, pause_template)
        if result:
            x, y, confidence = result
            h, w = pause_template.shape
            state, color = analyze_button_color(pil_image, x, y, w, h, button_type="pause")
            pause_state = state
            buttons.append(ButtonInfo(
                button_type=ButtonType.PAUSE,
                x=x, y=y, width=w, height=h,
                state=state, color_type=color, confidence=confidence
            ))
    
    # 查找停止按钮
    if stop_template is not None:
        result = find_button_by_template(gray_image, stop_template)
        if result:
            x, y, confidence = result
            h, w = stop_template.shape
            state, color = analyze_button_color(pil_image, x, y, w, h)
            stop_state = state
            stop_color = color
            buttons.append(ButtonInfo(
                button_type=ButtonType.STOP,
                x=x, y=y, width=w, height=h,
                state=state, color_type=color, confidence=confidence
            ))
    
    # 推断游戏状态
    game_state = infer_game_state(play_state, pause_state, stop_state, play_color, stop_color)
    
    return ToolbarState(
        play=play_state,
        pause=pause_state,
        stop=stop_state,
        game_state=game_state,
        buttons=buttons
    )


def infer_game_state(play: ButtonState, pause: ButtonState, stop: ButtonState,
                     play_color: str = None, stop_color: str = None) -> str:
    """
    根据按钮状态推断游戏状态
    
    规则（基于实际观察，按优先级）：
    1. 暂停按钮深灰色(enabled) -> 游戏运行中
    2. 暂停按钮浅灰色(disabled) -> 游戏停止
    3. 如果暂停按钮未找到，使用其他信号：
       - 停止按钮红色(enabled) -> 游戏运行中
       - 停止按钮灰色(disabled) -> 游戏停止
       - 播放按钮绿色(enabled) -> 游戏停止
       - 播放按钮红色(enabled) -> 游戏运行中（播放按钮变成恢复按钮）
    """
    # 优先使用暂停按钮状态
    if pause == ButtonState.ENABLED:
        return "running"
    elif pause == ButtonState.DISABLED:
        return "stopped"
    
    # 暂停按钮未找到时，使用其他信号
    # 停止按钮红色 = 游戏运行中
    if stop == ButtonState.ENABLED and stop_color == "red":
        return "running"
    
    # 停止按钮灰色 = 游戏停止
    if stop == ButtonState.DISABLED:
        return "stopped"
    
    # 播放按钮绿色 = 游戏停止
    if play == ButtonState.ENABLED and play_color == "green":
        return "stopped"
    
    # 播放按钮红色 = 游戏运行中（播放按钮变成恢复按钮，显示红色）
    if play == ButtonState.ENABLED and play_color == "red":
        return "running"
    
    # 最终默认
    return "stopped"


def detect_toolbar_state_with_debug(hwnd: int, debug_output_path: str = None) -> Tuple[Optional[ToolbarState], dict]:
    """
    检测工具栏状态（带调试信息）
    """
    debug_info = {}
    
    # 如果窗口最小化，先恢复
    restore_window_if_minimized(hwnd)
    
    # 截取窗口
    pil_image = capture_window_to_image(hwnd)
    if pil_image is None:
        debug_info["error"] = "Failed to capture window"
        return None, debug_info
    
    debug_info["window_size"] = pil_image.size
    
    # 转换为 OpenCV 格式
    cv_image = pil_to_cv2(pil_image)
    gray_image = cv2.cvtColor(cv_image, cv2.COLOR_BGR2GRAY)
    
    # 加载模板
    play_template = load_template("play")
    pause_template = load_template("pause")
    stop_template = load_template("stop")
    
    debug_info["templates_loaded"] = {
        "play": play_template is not None,
        "pause": pause_template is not None,
        "stop": stop_template is not None
    }
    
    buttons = []
    play_state = ButtonState.UNKNOWN
    pause_state = ButtonState.UNKNOWN
    stop_state = ButtonState.UNKNOWN
    play_color = None
    stop_color = None
    
    # 查找播放按钮
    if play_template is not None:
        result = find_button_by_template(gray_image, play_template)
        if result:
            x, y, confidence = result
            h, w = play_template.shape
            state, color = analyze_button_color(pil_image, x, y, w, h)
            play_state = state
            play_color = color
            buttons.append(ButtonInfo(
                button_type=ButtonType.PLAY,
                x=x, y=y, width=w, height=h,
                state=state, color_type=color, confidence=confidence
            ))
            debug_info["play"] = {"x": x, "y": y, "confidence": confidence, "state": state.value, "color": color}
        else:
            debug_info["play"] = {"error": "not found"}
    
    # 查找暂停按钮
    if pause_template is not None:
        result = find_button_by_template(gray_image, pause_template)
        if result:
            x, y, confidence = result
            h, w = pause_template.shape
            state, color = analyze_button_color(pil_image, x, y, w, h, button_type="pause")
            pause_state = state
            buttons.append(ButtonInfo(
                button_type=ButtonType.PAUSE,
                x=x, y=y, width=w, height=h,
                state=state, color_type=color, confidence=confidence
            ))
            debug_info["pause"] = {"x": x, "y": y, "confidence": confidence, "state": state.value, "color": color}
        else:
            debug_info["pause"] = {"error": "not found"}
    
    # 查找停止按钮
    if stop_template is not None:
        result = find_button_by_template(gray_image, stop_template)
        if result:
            x, y, confidence = result
            h, w = stop_template.shape
            state, color = analyze_button_color(pil_image, x, y, w, h)
            stop_state = state
            stop_color = color
            buttons.append(ButtonInfo(
                button_type=ButtonType.STOP,
                x=x, y=y, width=w, height=h,
                state=state, color_type=color, confidence=confidence
            ))
            debug_info["stop"] = {"x": x, "y": y, "confidence": confidence, "state": state.value, "color": color}
        else:
            debug_info["stop"] = {"error": "not found"}
    
    # 保存调试图像
    if debug_output_path:
        debug_img = cv_image.copy()
        colors = {
            ButtonType.PLAY: (0, 255, 0),    # 绿色
            ButtonType.PAUSE: (255, 255, 0),  # 青色
            ButtonType.STOP: (0, 0, 255)      # 红色
        }
        for btn in buttons:
            color = colors.get(btn.button_type, (255, 255, 255))
            cv2.rectangle(debug_img, (btn.x, btn.y), 
                         (btn.x + btn.width, btn.y + btn.height), color, 2)
            label = f"{btn.button_type.value}:{btn.color_type}"
            cv2.putText(debug_img, label, (btn.x, btn.y - 5), 
                       cv2.FONT_HERSHEY_SIMPLEX, 0.5, color, 1)
        cv2.imwrite(debug_output_path, debug_img)
        debug_info["debug_image"] = debug_output_path
    
    # 推断游戏状态
    game_state = infer_game_state(play_state, pause_state, stop_state, play_color, stop_color)
    debug_info["game_state"] = game_state
    
    toolbar_state = ToolbarState(
        play=play_state,
        pause=pause_state,
        stop=stop_state,
        game_state=game_state,
        buttons=buttons
    )
    
    return toolbar_state, debug_info

File: src/toolbar-detector.mjs
import { existsSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import sharp from "sharp";

const __dirname = dirname(fileURLToPath(import.meta.url));
const TEMPLATE_DIR = join(__dirname, "..", "templates");
const THEME_DIRS = ["dark", "light"]; // 优先尝试 dark 主题

const { cv } = await import("opencv-wasm");

function rgbToHsv(r, g, b) {
  r /= 255;
  g /= 255;
  b /= 255;
  const max = Math.max(r, g, b);
  const min = Math.min(r, g, b);
  const diff = max - min;
  const v = max * 100;
  const s = max === 0 ? 0 : (diff / max) * 100;
  let h = 0;
  if (diff !== 0) {
    if (max === r) h = 60 * (((g - b) / diff) % 6);
    else if (max === g) h = 60 * ((b - r) / diff + 2);
    else h = 60 * ((r - g) / diff + 4);
  }
  if (h < 0) h += 360;
  return [h, s, v];
}

async function loadTemplateGray(name) {
  // 尝试从各主题目录加载模板，返回所有找到的模板
  const templates = [];
  
  for (const theme of THEME_DIRS) {
    const p = join(TEMPLATE_DIR, theme, `${name}.png`);
    if (!existsSync(p)) continue;
    const { data, info } = await sharp(p).grayscale().raw().toBuffer({ resolveWithObject: true });
    templates.push({ data, width: info.width, height: info.height, theme });
  }
  
  // 兼容旧目录结构（直接在 templates/ 下）
  const legacyPath = join(TEMPLATE_DIR, `${name}.png`);
  if (existsSync(legacyPath)) {
    const { data, info } = await sharp(legacyPath).grayscale().raw().toBuffer({ resolveWithObject: true });
    templates.push({ data, width: info.width, height: info.height, theme: "legacy" });
  }
  
  return templates.length > 0 ? templates : null;
}

function bufferToMat(data, width, height, channels) {
  const type = channels === 1 ? cv.CV_8UC1 : cv.CV_8UC3;
  const mat = new cv.Mat(height, width, type);
  mat.data.set(data);
  return mat;
}

function findButtonByTemplate(screenshotGray, template, threshold = 0.6) {
  if (!template) return null;
  const result = new cv.Mat();
  cv.matchTemplate(screenshotGray, template, result, cv.TM_CCOEFF_NORMED);
  const minMax = cv.minMaxLoc(result);
  result.delete();
  if (minMax.maxVal >= threshold) {
    return { x: minMax.maxLoc.x, y: minMax.maxLoc.y, confidence: minMax.maxVal };
  }
  return null;
}

// 从多个模板中找到最佳匹配
function findBestMatch(screenshotGray, templates, threshold = 0.55) {
  if (!templates || templates.length === 0) return null;
  
  let bestMatch = null;
  let bestTemplate = null;
  
  for (const tpl of templates) {
    const tplMat = bufferToMat(tpl.data, tpl.width, tpl.height, 1);
    const match = findButtonByTemplate(screenshotGray, tplMat, threshold);
    tplMat.delete();
    
    if (match && (!bestMatch || match.confidence > bestMatch.confidence)) {
      bestMatch = match;
      bestTemplate = tpl;
    }
  }
  
  return bestMatch ? { match: bestMatch, template: bestTemplate } : null;
}

function analyzeButtonColor(rgbData, imgWidth, x, y, w, h, buttonType = null) {
  const COLOR_THRESHOLD = 50;
  let redCount = 0, greenCount = 0, blueCount = 0;
  let darkPixelCount = 0, totalDarkBrightness = 0;
  let brightPixelCount = 0;
  let total = 0;

  for (let dy = 0; dy < h; dy++) {
    for (let dx = 0; dx < w; dx++) {
      const px = x + dx;
      const py = y + dy;
      if (px < 0 || py < 0 || px >= imgWidth) continue;
      const idx = (py * imgWidth + px) * 3;
      const r = rgbData[idx], g = rgbData[idx + 1], b = rgbData[idx + 2];
      if (r === undefined) continue;
      const [hue, sat, val] = rgbToHsv(r, g, b);
      const brightness = (r + g + b) / 3;
      total++;

      if (brightness < 200) {
        darkPixelCount++;
        totalDarkBrightness += brightness;
      }
      if (brightness > 120) {
        brightPixelCount++;
      }

      if (sat > COLOR_THRESHOLD && val > 30) {
        if (hue <= 30 || hue >= 330) redCount++;
        else if (hue >= 80 && hue <= 160) greenCount++;
        else if (hue >= 180 && hue <= 250) blueCount++;
      }
    }
  }

  if (total === 0) return ["unknown", "unknown"];

  if (buttonType === "pause" && darkPixelCount > 0) {
    const avgDark = totalDarkBrightness / darkPixelCount;
    const brightRatio = brightPixelCount / total;
    // Windows light theme: stopped avgDark ~154, running avgDark ~53
    if (avgDark > 120) return ["disabled", "light"];
    // Mac dark theme: running brightRatio ~0.275, stopped ~0.167
    if (brightRatio > 0.22) return ["enabled", "dark"];
    // Low brightRatio + low avgDark = dark theme disabled
    return ["disabled", "dark"];
  }

  const minRatio = 0.03;
  if (redCount > total * minRatio) {
    // For stop button in Mac dark theme, distinguish bright red (enabled) from dim red (disabled)
    if (buttonType === "stop" && darkPixelCount > 0) {
      const avgDark = totalDarkBrightness / darkPixelCount;
      if (avgDark < 50) return ["disabled", "red"];
    }
    return ["enabled", "red"];
  }
  if (greenCount > total * minRatio) return ["enabled", "green"];
  if (blueCount > total * minRatio) return ["enabled", "blue"];
  return ["disabled", "gray"];
}

function inferGameState(play, pause, stop, playColor, stopColor) {
  if (pause === "enabled") return "running";
  if (pause === "disabled") return "stopped";
  // play 红色优先判断 running（比 stop disabled 更可靠）
  if (play === "enabled" && playColor === "red") return "running";
  if (play === "enabled" && playColor === "green") return "stopped";
  if (stop === "enabled" && stopColor === "red") return "running";
  if (stop === "disabled") return "stopped";
  return "stopped";
}

export async function detectToolbarState(captureResult) {
  if (!captureResult) return null;

  const { data: rgbData, width, height } = captureResult;

  const grayBuf = await sharp(rgbData, { raw: { width, height, channels: 3 } })
    .grayscale()
    .raw()
    .toBuffer();

  const grayMat = bufferToMat(grayBuf, width, height, 1);

  const playTpls = await loadTemplateGray("play");
  const pauseTpls = await loadTemplateGray("pause");
  const stopTpls = await loadTemplateGray("stop");

  // 按主题分组匹配，选择最佳主题
  const themeResults = {};
  
  for (const theme of [...THEME_DIRS, "legacy"]) {
    const buttons = [];
    
    for (const [name, tpls, type] of [
      ["play", playTpls, null],
      ["pause", pauseTpls, "pause"],
      ["stop", stopTpls, null],
    ]) {
      if (!tpls) continue;
      
      // 只使用当前主题的模板
      const themeTpls = tpls.filter(t => t.theme === theme);
      if (themeTpls.length === 0) continue;
      
      const result = findBestMatch(grayMat, themeTpls);
      if (result) {
        const { match, template } = result;
        const [state, color] = analyzeButtonColor(
          rgbData, width, match.x, match.y, template.width, template.height, type,
        );
        buttons.push({
          type: name,
          x: match.x, y: match.y,
          width: template.width, height: template.height,
          state, colorType: color, confidence: match.confidence,
          theme: template.theme,
        });
      }
    }
    
    if (buttons.length > 0) {
      // 计算主题得分：匹配数量 + 平均置信度 + 位置一致性
      const avgConfidence = buttons.reduce((s, b) => s + b.confidence, 0) / buttons.length;
      const yValues = buttons.map(b => b.y);
      const yConsistency = yValues.length > 1 ? 
        (Math.max(...yValues) - Math.min(...yValues) < 50 ? 0.2 : 0) : 0.1;
      
      themeResults[theme] = {
        buttons,
        score: buttons.length * 0.3 + avgConfidence + yConsistency,
      };
    }
  }

  grayMat.delete();

  // 选择得分最高的主题
  let bestTheme = null;
  let bestScore = 0;
  for (const [theme, result] of Object.entries(themeResults)) {
    if (result.score > bestScore) {
      bestScore = result.score;
      bestTheme = theme;
    }
  }

  if (!bestTheme) {
    return {
      play: "unknown",
      pause: "unknown",
      stop: "unknown",
      gameState: "stopped",
      buttons: [],
      theme: null,
    };
  }

  const buttons = themeResults[bestTheme].buttons;
  let playState = "unknown", pauseState = "unknown", stopState = "unknown";
  let playColor = null, stopColor = null;

  for (const btn of buttons) {
    if (btn.type === "play") { playState = btn.state; playColor = btn.colorType; }
    if (btn.type === "pause") { pauseState = btn.state; }
    if (btn.type === "stop") { stopState = btn.state; stopColor = btn.colorType; }
  }

  const gameState = inferGameState(playState, pauseState, stopState, playColor, stopColor);

  return {
    play: playState,
    pause: pauseState,
    stop: stopState,
    gameState,
    buttons,
    theme: bestTheme,
  };
}

export async function detectToolbarStateFromFile(imagePath) {
  const { data, info } = await sharp(imagePath)
    .removeAlpha()
    .raw()
    .toBuffer({ resolveWithObject: true });
  return detectToolbarState({ data, width: info.width, height: info.height });
}

export { rgbToHsv, analyzeButtonColor, inferGameState };
